sdk: add a yml example on how to use fdd to create container channels

samoht opened the pull request:
Lots of boilerplate for now on, will work on upstreaming that in the tool
properly if needed later.

samoht commented on projects/miragesdk/examples/fdd.yml:
```diff
@@ -0,0 +1,40 @@
+files:
+  - path: etc/init.d/020-fdd
+    mode: "0700"
+    contents: "/bin/fdd init -vv -s /var/run/fdd.sock"
```
Actually it's a bit overkilling to use `onboot` service to create the channels, maybe it should only be in the `fdd` init script.

justincormack commented on projects/miragesdk/examples/fdd.yml:
```diff
@@ -0,0 +1,40 @@
+    image: "linuxkit/sysctl:3aa6bc663c2849ef239be7d941d3eaf3e6fcc018"
+  - name: channel-net-eng
+    image: samoht/fdd
+    binds: [ "/var/run/fdd.sock:/var/run/fdd.sock" ]
```
please don't use `/var/run` its just `/run`, `/var/run` is just a symlink.

samoht commented:
I've fixed the yaml example. I also added a `fdd exec` to wrap command in an environment where file-descriptor are pre-open. For instance, this now works:

```
$ fdd init &
$ fdd share /tmp/foo
$ fdd exec -m /tmp/foo:1 -- echo hello world!
$ fdd exec -m /tmp/foo:0 -- cat
hello world!
```

samoht commented:
At the moment, it's just wrapper script to be able to prototype stuff. I don't expect end user to see anything of this -- I plan to push some of these stuff to the moby tool (for instance, use a new `channel` section instead of having to modify init scripts) and I will experiment by changing `container-shim` instead of having to call `fdd exec`.

Which part of the current design do you find ugly?

justincormack commented:
The whole point of anonymous socketpairs is to control exactly which process gets the two ends of the channel, and you totally lose this if you just get it via the filesystem instead, when some other process could race to get the other file descriptor.

samoht commented:
But if you control the init scripts and the packaging, you can decide which process can see the unix domain socket (via bind mounts for instance) and in which order. Do you have a better idea?

The good thing about the current approach is that it's easy enough to test that all this (rather complex) is working..

samoht commented:
e.g. it's easy to extend to something like that:

```
 $ fdd init &
$ fdd share /tmp/foo --secret foo
$ fdd exec -m /tmp/foo:1 --secret foo -- echo hello world!
$ fdd exec -m /tmp/foo:0 -- cat
<error>
$ fdd exec -m /tmp/foo:0 --secret foo -- cat
hello world!
```

justincormack commented on projects/miragesdk/examples/fdd.yml:
```diff
@@ -0,0 +1,34 @@
+    contents: "/bin/fdd init -vv -s /run/fdd.sock"
+  - path: etc/init.d/030-fdd-share
+    mode: "0700"
+    contents: |
```
these files need to start with `#!/bin/sh`

samoht replied:
thx, force-pushed a fix

justincormack commented:
Thats no better than just having one program listen on a unix socket in the filesystem while the other one connects to it, but much more complicated.

samoht commented:
Yes, this is indeed more complicated but because I am trying to limit the number of syscalls and the number of operations that the 2 processes have to do on startup.

If you have 2 processes and A needs to listens and accept on `/tmp/foo` and B connects on `/tmp/foo`. A needs to be started before B and/or B needs to wait for A to be ready to accept connections. A and B also needs to both see the same unix domain socket in their namespace.

With `fdd`, A and B can just read/write on the given file-descriptor (for instance fd 3), with:
```
fd exec /tmp/foo:3 -- ./A
fd exec /tmp/foo:3 -- ./B
```

samoht commented:
Also this let change the way the two side of the socketpair are created, for instance using a shared pages + ring buffer, or network connection, etc. E.g. the process won't care as it will still read from fd 3, but the init scripts (done by `fdd` or something similar) could handle a more complex connection startup logic.

justincormack commented:
Going to merge as just a project example.
